### PR TITLE
Rework to avoid dangling references to OpSpec instances

### PR DESCRIPTION
**Pull request by jantonguirao**
Signed-off-by: Joaquin Anton <janton@nvidia.com>

## Description
- [x] **Bug fix** (*non-breaking change which fixes an issue*)
- [ ] **New feature** (*non-breaking change which adds functionality*)
- [ ] **Breaking change** (*fix or feature that would cause existing functionality to not work as expected*)
- [ ] **Refactoring** (*Redesign of existing code that doesn't affect functionality*)
- [ ] **Other** (*e.g. Documentation, Tests, Configuration*)

### What happened in this PR

Fixes a bug observed when de-serializing pipelines. The issue seems to be linked to holding references to OpSpec instances that are stack variables. This PR reworks the way that OpSpecs instances are passed around. To make it more obvious, const pointers to OpSpec are passed to utility classes. 

#### Additional information
- Affected modules and functionalities:

Erase, Slice based ops, RandomBBoxCrop, Warp, COCOLoader

- Key points relevant for the review:

NA

### Tests
- [ ] Existing tests apply
- [ ] New tests added
  - [ ] Python tests
  - [ ] GTests
  - [ ] Benchmark
  - [ ] Other
- [x] N/A

### Documentation
- [ ] Existing documentation applies
- [ ] Documentation updated
  - [ ] Docstring
  - [ ] Doxygen
  - [ ] RST
  - [ ] Jupyter
  - [ ] Other
- [x] N/A

#### Requirements
- [ ] Implements new requirements
- [ ] Affects existing requirements
- [x] N/A

**REQ IDs**: N/A

**JIRA TASK**: N/A

### CODE REVIEW

File: dali/operators/generic/erase/erase.cu [hunk 1]
```diff
@@ -99,7 +99,7 @@ class EraseImplGpu : public OpImplBase<GPUBackend> {
   }
 
  private:
-  const OpSpec &spec_;
+  const OpSpec spec_;
```
Comment (JanuszL) [inline dali/operators/generic/erase/erase.cu:102]: Does it make sense to add a test for that?

Comment (jantonguirao) [reply dali/operators/generic/erase/erase.cu:102]: Do you have a good idea how to cover all ops without having to write a lot of code? The issue happens during run.

Comment (JanuszL) [reply dali/operators/generic/erase/erase.cu:102]: Not really. If you don't have any let us skip it.

Comment (mzient) [reply dali/operators/generic/erase/erase.cu:102]: Please revert. This class is safe - it's a reference to a member of the enclosing object - the lifetime of the real object is longer than that of this class.

File: dali/operators/image/crop/bbox_crop.h [hunk 2]
```diff
@@ -36,7 +36,7 @@ class RandomBBoxCrop : public Operator<Backend> {
  private:
   std::unique_ptr<OpImplBase<Backend>> impl_;
   int impl_ndim_ = -1;
-  OpSpec spec_;
+  using Operator<Backend>::spec_;
```
Comment (mzient) [inline dali/operators/image/crop/bbox_crop.h:39]: :+1:  That's the correct way.

File: dali/operators/reader/loader/coco_loader.h [hunk 3]
```diff
@@ -200,7 +200,7 @@ class DLL_PUBLIC CocoLoader : public FileLabelLoader {
   void SavePreprocessedAnnotations(const std::string &path, const ImageIdPairs &image_id_pairs);
 
  private:
-  const OpSpec &spec_;
+  const OpSpec spec_;
```
Comment (jantonguirao) [inline dali/operators/reader/loader/coco_loader.h:203]: That's the only extra copy now. The code would be too brittle or messy without it.

**Review by JanuszL — APPROVED** [review]

File: dali/operators/image/convolution/gaussian_blur.cc [hunk 4]
```diff
@@ -121,8 +121,12 @@ class GaussianBlurOpCpu : public OpImplBase<CPUBackend> {
   using Kernel = kernels::SeparableConvolutionCpu<Out, In, float, axes, has_channels>;
   static constexpr int ndim = Kernel::ndim;
 
-  explicit GaussianBlurOpCpu(const OpSpec& spec, const DimDesc& dim_desc)
-      : spec_(spec), dim_desc_(dim_desc) {}
+  /**
+   * @param spec  Pointer to a persistent OpSpec object,
+   *              which is guaranteed to be alive for the entire lifetime of this object
+   */
+  explicit GaussianBlurOpCpu(const OpSpec* spec, const DimDesc& dim_desc)
```
Comment (klecki) [inline dali/operators/image/convolution/gaussian_blur.cc:128]: Why change it to pointer just to take address at the call site and dereference it here?

Comment (mzient) [reply dali/operators/image/convolution/gaussian_blur.cc:128]: See the comment :)

**Review by mzient — APPROVED** [review]

**Comment by mzient** [comment]
!build

File: dali/operators/image/convolution/gaussian_blur.cc (hunk 4 above)
Comment (klecki) [reply dali/operators/image/convolution/gaussian_blur.cc:128]: That's basically a definition of what pass by reference means.

Comment (jantonguirao) [reply dali/operators/image/convolution/gaussian_blur.cc:128]: Imagine:

```
GaussianBlurOpCpu impl_;
ParentOp(const OpSpec &spec, ...) 
  : impl_ (spec) {}

// in a function
OpSpec spec;
...
return ParentOp(spec);
```
^^ Looks legit from here, but it's actually storing a reference to a stack variable

```
GaussianBlurOpCpu impl_;
ParentOp(const OpSpec &spec, ...) 
  : impl_ (&spec) {}
 ```
^^ If I have to write here, it becomes obvious that I am doing something wrong, taking the address of a ctor argument.



```
GaussianBlurOpCpu impl_;
ParentOp(const OpSpec &spec, ...) 
  : spec_(spec), impl_ (&spec_) {}
 ```
 ^^ And would make it more obvious that I should pass a pointer to a persistent object

Comment (jantonguirao) [reply dali/operators/image/convolution/gaussian_blur.cc:128]: In other words, I changed the internal impl to take a pointer to make it obvious that it's storing a view of the OpSpec only

**Pull request merged**